documentation/docs/guides/typing.md

```diff
@@ -116,3 +126,77 @@ export type Logger = {
 ```
 
 If we tried to do `deps.logger.notALogMethod(...)` we would expect a type error.  Fusion.js will automatically hook up these injected types when they are available, ensuring more seamless type safety across the Fusion.js ecosystem.
+
+#### Universal plugin types
+
+It's possible for plugins to be non-trivial enough that they have similar but not identical type signatures on server and browser.
+
+The recommended way to type those types of plugins is to create separate plugins and tokens for each environment:
```

Comment:
For what it's worth, we don't follow this advice internally.  There are a few places where we use [the same types across environments](https://github.com/fusionjs/fusion-plugin-i18n/blob/master/src/types.js#L19) when the actual types differ.

This is (1) partially due to retrofitting types after the fact, and (2) to avoid introducing new environment-specific tokens for the same plugins.

Reply:
Yean, I'm aware that we basically lie for our own plugins, but I think for new development, we should discourage people from following that practice. Thoughts?

Reply:
I think they'd be different problems, but I'm more on board w/ the issues of having two separate tokens for server/browser registrations. In that case, at best, you get an error saying "Such and such token was not registered" and at worst, the runtime simply does nothing because you forgot to add the code to do it.

With a single token overloading both server and client type, there's no way to correctly type either server-side or browser-side usage. At best you need to add type checks everywhere to disambiguate between `A | B`, and at worst you get uncaught runtime type errors because you used a part of the server type signature in the browser or vice-versa.

Reply:
Agreed that from a static typing perspective, different tokens makes our lives easier as it avoids the need for any hacks or type refinements.

I don't think we have (or feasibly can) address the issue of the client and server implementations having an implicit dependency upon each other that isn't codified anywhere.  We could not get any error handling around this if done today to say "you are missing the server implementation for this plugin" for example.  So it would be the worst case from above - runtime does not work as expected, since only half of the plugin is working.

That being said, that trade-off is possibly sufficient if we are looking to avoid any compiler hacks (e.g. utility functions to hide this) and want a cleaner static typing experience.  I'd be interested to see what other folks' opinions are here, and whether we should adopt this approach for our plugins as well.

Reply:
Also just a little concerned that we would give this advice if we don't follow it ourselves. If we are going to document this approach, we should probably update our existing plugins to as well. It might be worth making a list of migrations we would have to do to follow this?